File: sdk/src/core/spanBatcher.js
/**
 * Span Batcher for OTLP/HTTP ingest
 * Collects spans and flushes to /api/otlp/v1/traces
 */

class SpanBatcher {
  constructor(client, batchSize = 100, flushIntervalMs = 10000) {
    this.client = client;
    this.batchSize = batchSize;
    this.flushIntervalMs = flushIntervalMs;
    this.queue = [];
    this.flushTimer = null;
    this.isShuttingDown = false;
  }

  /**
   * Add a span to the queue
   */
  enqueue(span) {
    if (this.isShuttingDown) {
      return;
    }

    this.queue.push(span);

    // Auto-flush if queue exceeds batchSize
    if (this.queue.length >= this.batchSize) {
      this.flush();
    }
  }

  /**
   * Start background flush timer
   */
  start() {
    if (this.flushTimer) {
      return;
    }

    this.flushTimer = setInterval(() => {
      if (this.queue.length > 0) {
        this.flush();
      }
    }, this.flushIntervalMs);
    // Don't keep the event loop alive solely for the flush timer —
    // callers that want the process to linger should use sdk.start()
    // + sdk.shutdown() explicitly. Mirrors the event batcher's behavior.
    if (typeof this.flushTimer?.unref === 'function') {
      this.flushTimer.unref();
    }
  }

  /**
   * Stop background flush timer
   */
  stop() {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
    }
  }

  /**
   * Flush all queued spans to OTLP endpoint
   */
  async flush() {
    if (this.queue.length === 0) {
      return;
    }

    const batch = this.queue.splice(0, this.batchSize);
    const payload = this._buildOtlpPayload(batch);

    try {
      const response = await fetch(`${this.client.endpoint}/api/otlp/v1/traces`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-SDK-Key': this.client.apiKey,
        },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`HTTP ${response.status}: ${error}`);
      }

      this.client._log(`[Span Flush] Success (${batch.length} spans)`);
    } catch (error) {
      this.client._log('[ERROR] Span flush failed:', error.message);
      // Re-queue spans on failure
      this.queue.unshift(...batch);
    }
  }

  /**
   * Build OTLP/HTTP payload from spans
   */
  _buildOtlpPayload(spans) {
    const otlpSpans = spans.map(s => s.toOtlpSpan());

    return {
      resourceSpans: [
        {
          resource: {
            attributes: [
              { key: 'service.name', value: { stringValue: this.client.appName } },
              { key: 'service.version', value: { stringValue: '1.0.0' } },
            ],
          },
          scopeSpans: [
            {
              scope: {
                name: '@fathom/sdk',
                version: '1.0.0',
              },
              spans: otlpSpans,
            },
          ],
        },
      ],
    };
  }

  /**
   * Shutdown: flush remaining and stop timer
   */
  async shutdown() {
    this.isShuttingDown = true;
    this.stop();
    if (this.queue.length > 0) {
      await this.flush();
    }
  }
}

export { SpanBatcher };

File: sdk/python/src/fathom_sdk/span_batcher.py
"""Span Batcher for OTLP/HTTP ingest."""

import asyncio
import json
from typing import List, Optional
from .tracer import Span


class SpanBatcher:
    """Collects spans and flushes to /api/otlp/v1/traces."""

    def __init__(
        self,
        client,
        batch_size: int = 100,
        flush_interval_seconds: float = 10.0,
        debug: bool = False,
    ):
        """Initialize span batcher."""
        self.client = client
        self.batch_size = batch_size
        self.flush_interval_seconds = flush_interval_seconds
        self.debug = debug
        self.queue: List[Span] = []
        self.flush_task: Optional[asyncio.Task] = None
        self.is_shutting_down = False

    def enqueue(self, span: Span) -> None:
        """Add a span to the queue."""
        if self.is_shutting_down:
            return

        self.queue.append(span)

        # Auto-flush if queue exceeds batchSize
        if len(self.queue) >= self.batch_size:
            asyncio.create_task(self.flush())

    async def flush(self) -> None:
        """Flush all queued spans to OTLP endpoint."""
        if len(self.queue) == 0:
            return

        batch = self.queue[:self.batch_size]
        self.queue = self.queue[self.batch_size:]

        payload = self._build_otlp_payload(batch)

        try:
            import httpx
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{self.client.endpoint}/api/otlp/v1/traces",
                    json=payload,
                    headers={
                        'Content-Type': 'application/json',
                        'X-SDK-Key': self.client.api_key,
                    },
                    timeout=30.0,
                )

                if response.status_code != 200:
                    raise Exception(f"HTTP {response.status_code}: {response.text}")

                self._log(f"[Span Flush] Success ({len(batch)} spans)")
        except Exception as error:
            self._log(f"[ERROR] Span flush failed: {error}")
            # Re-queue spans on failure
            self.queue = batch + self.queue

    def _build_otlp_payload(self, spans: List[Span]) -> dict:
        """Build OTLP/HTTP payload from spans."""
        otlp_spans = [s.to_otlp_span() for s in spans]

        return {
            'resourceSpans': [
                {
                    'resource': {
                        'attributes': [
                            {'key': 'service.name', 'value': {'stringValue': self.client.app_name}},
                            {'key': 'service.version', 'value': {'stringValue': '1.0.0'}},
                        ],
                    },
                    'scopeSpans': [
                        {
                            'scope': {
                                'name': '@fathom/sdk',
                                'version': '1.0.0',
                            },
                            'spans': otlp_spans,
                        },
                    ],
                },
            ],
        }

    async def shutdown(self) -> None:
        """Shutdown: flush remaining and cleanup."""
        self.is_shutting_down = True
        if len(self.queue) > 0:
            await self.flush()

    def _log(self, msg: str) -> None:
        """Debug logging."""
        if self.debug:
            print(f"[FATHOM] {msg}")

    def size(self) -> int:
        """Get current queue size."""
        return len(self.queue)
